compositor_common/src/scene/builtin_transformations.rs

```diff
@@ -105,15 +108,48 @@ impl FromStr for MirrorMode {
 }
 
 impl BuiltinSpec {
-    pub fn validate(&self, node_spec: &NodeSpec) -> Result<(), BuiltinSpecValidationError> {
+    fn transformation_name(&self) -> String {
         match self {
-            BuiltinSpec::TransformToResolution { .. } => {
-                if node_spec.input_pads.len() != 1 {
-                    return Err(BuiltinSpecValidationError::TransformToResolutionExactlyOneInput);
-                }
+            BuiltinSpec::TransformToResolution { .. } => "transform_to_resolution".to_owned(),
+            BuiltinSpec::FixedPositionLayout { .. } => "fixed_position_layout".to_owned(),
+            BuiltinSpec::TiledLayout { .. } => "tiled_layout".to_owned(),
+            BuiltinSpec::MirrorImage { .. } => "mirror_image".to_owned(),
+            BuiltinSpec::CornersRounding { .. } => "corners_rounding".to_owned(),
+        }
+    }
 
-                Ok(())
-            }
+    fn valid_inputs_count(&self) -> ValidNodeInputsCount {
```

Comment:
This is a good solution for input count but it does not scale. I'm fine if you want to lond it, but this will need to be reworked to implement constraint, so I don't know if it makes sense to take this approach.

Also I don't think this should be limited only to builtin types (although it most useful here)

Reply:
I thought about returning the struct holding all node constraints and then checking. This would need to take place in the renderer, during scene update (to have access to user-defined constraints of shaders/web renderers). I can't come up with constraints different than input count and static/dynamic resolution. I think implementing static/dynamic resolution validation is a wrong idea because it makes API more difficult for newcomers. 

Can you think of any other potential constraints? I don't want to unnecessarily overcomplicate it.

Reply:
> I think implementing static/dynamic resolution validation is a wrong idea because it makes API more difficult for newcomers.

This is an argument to not have those constraints on our components, but the user might want to add constraints like this on their own shaders to protect them from mistakes.

Potential constraints:
- input count
  - exact value (0, 1, n)
  - less than n
  - at least n
  - has to be one of the specific values
- resolution
  - exact resolution
  - constant resolution
  - max input size (either on specific dimension or overall)
- static content (only allow inputs that are static and need to be rendered only once)
- infallible (always produces an output either via fallback or on its own)
- You can't pass the same input more than once (not sure what's the use case)
- Allow only nodes of specific type e.g. only inputs/ only images /...
- Allow only inputs named according to specific naming conventions e.g. via regexp or some prefix/suffix

Reply:
Ok, I'll add shaders and web render inputs count specification, make builtins return validation struct, and validate it in renderer on scene update.
Having some validation in compositor common makes sense to me in transformations, e.g. `FixedPositionLayout`, where each input requires some params (we can provide more better error message). Other checks should be uniform across all transformations.

Other constraints are outside of the scope of this PR.

Reply:
> I'll add shaders and web render inputs count specification

I think it's outside of scope of this PR too. Unless you want to do this.

I would just create a Constraint abstraction attach it to the node (e.g. method that returns constraints) and run constraint validation on scene update.